internal/certificates/certificates.go

```diff
@@ -0,0 +1,146 @@
+package certificates
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"errors"
+	"math/big"
+	"net"
+	"time"
+)
+
+// GenerateCACertificate generates a one time discardable root CA for the running GoCert webserver
+func GenerateCACertificate() (string, string, error) {
+	caCert := &x509.Certificate{
+		SerialNumber: big.NewInt(2019),
+		Subject: pkix.Name{
+			Organization: []string{"Canonical, INC."},
+			Country:      []string{"US"},
+		},
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().AddDate(10, 0, 0),
+		IsCA:                  true,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+	}
+	caPrivateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return "", "", err
+	}
+
+	caBytes, err := x509.CreateCertificate(rand.Reader, caCert, caCert, &caPrivateKey.PublicKey, caPrivateKey)
+	if err != nil {
+		return "", "", err
+	}
+
+	caCertPEM := new(bytes.Buffer)
+	err = pem.Encode(caCertPEM, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: caBytes,
+	})
+	if err != nil {
+		return "", "", err
+	}
+
+	caPrivateKeyPEM := new(bytes.Buffer)
+	err = pem.Encode(caPrivateKeyPEM, &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(caPrivateKey),
+	})
+	if err != nil {
+		return "", "", err
+	}
+	return caCertPEM.String(), caPrivateKeyPEM.String(), nil
+}
+
+// GenerateSelfSignedCertificate will create a certificate and pk for GoCert coming from a rootCA
+// This certificate and PK is not saved anywhere.
+// This certificate and PK should either be saved somewhere, or a real certificate should be provided to GoCert
+func GenerateSelfSignedCertificate(caCertPEM, caPrivateKeyPEM string) (string, string, error) {
+	caCert, err := ParseCertificate(caCertPEM)
+	if err != nil {
```

Comment:
Right now, the only usage we make of the CA certificate is to generate a self-signed certificate. I do not think we should convert the return values to string to parse them again in the next function.

In fact, I think we should mostly not convert to/from strings. When it becomes necessary, it should happen at the edges of the program.

Reply:
This part is gone anyway but agreed